private/buf/bufformat/formatter.go

```diff
@@ -0,0 +1,921 @@
+// Copyright 2020-2022 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bufformat
+
+import (
+	"fmt"
+	"io"
+	"sort"
+	"strconv"
+	"strings"
+
+	"github.com/jhump/protocompile/ast"
+)
+
+// formatter writes an *ast.FileNode as a .proto file.
+type formatter struct {
+	writer   io.Writer
+	fileNode *ast.FileNode
+	indent   int
+}
+
+// newFormatter returns a new formatter for the given file.
+func newFormatter(writer io.Writer, fileNode *ast.FileNode) *formatter {
+	return &formatter{
+		writer:   writer,
+		fileNode: fileNode,
+	}
+}
+
+// Run runs the formatter and writes the file's content to the formatter's writer.
+func (f *formatter) Run() error {
+	return f.writeFile(f.fileNode)
+}
+
+// P prints a line to the generated output.
+func (f *formatter) P(elements ...interface{}) {
+	if len(elements) > 0 {
+		// Don't use an indent if we're just writing a newline.
+		fmt.Fprint(f.writer, strings.Repeat("  ", f.indent))
```

Comment:
I'm surprised this isn't caught by `errcheck`, i.e. I think you have to do `_, _ =`